test/CoreTypechecker.hs

```diff
@@ -64,19 +64,21 @@ depIdentity =
         )
     )
 
-{- TODO
-depIdentityCompTy ∷ AllAnnotation
+depIdentityCompTy ∷ ∀ primTy primVal. IR.Annotation primTy primVal
 depIdentityCompTy =
   ( SNat 0,
     IR.VPi
       (SNat 0)
-      (IR.VStar 0)
-      ( IR.vapp
-          All.all
-          (IR.VLam (IR.VPrimTy))
+      (IR.VStar 0) -- first input is of type type
+      ( const
+          ( IR.VPi
+              (SNat 1)
+              (IR.VNeutral (IR.NFree (IR.Local 0))) -- second input is of type of the first input
+              (const (IR.VNeutral (IR.NFree (IR.Local 0)))) -- the function's return is of the type that's the same as the second input
```

Comment:
I understand the code but not the comment. This is just the type we bound in the first input, right?

Reply:
Yes, it is of the type of the first input. But it's not the first input, because the first input is the type's type, i.e., `* 0`. 
 
The dependent identity is 

`\t.\x.x:t` 

the first input is `t`, the type of `x`, the second input is `x`, of type `t`. The return is `x`, of type `t`.

So the return type is the second input. 

Did I answer your question?

Reply:
Maybe I'm just mixing up the indices, but I'm still confused.

Per my counting:

`\t.\x:t.x:t` - `t` is the first input (a type, with value `t` and type `* 0` itself), `x` is the second input (with value `x` and type `t`), and the return *value* is `x`, which is of type `t`, which is the first input.

Pi should bind the *value* of the first input - which is the type `t` - not the *type* of the first input `* 0` (we don't care about that, I don't think it should be in scope at all).

Reply:
> Maybe I'm just mixing up the indices, but I'm still confused.
> 

The bound variables are in de bruijn indices, so in line 76, the index is 0 because `t` is the input bounded immediately before. In line 77, the input immediately before is `x:t` so again the index is 0.

> Per my counting:
> 
> `\t.\x:t.x:t` - `t` is the first input (a type, with value `t` and type `* 0` itself), `x` is the second input (with value `x` and type `t`), and the return _value_ is `x`, which is of type `t`, which (`t`) is the value of the first input.
> 
Yes that's the idea.

> Pi should bind the _value_ of the first input - which is the type `t` - not the _type_ of the first input `* 0` (we don't care about that, I don't think it should be in scope at all).

This is the annotation for the dependent identity. So these are all types.

Am I making sense?

Reply:
They are all types, yes, but whenever we type-check a dependent (pi) type, de-Bruijin index 0 should be the *value* of the function argument, right?

Then this pi type seems like it's returning a value of type `x`, which is not what we want - we want to return a value of type `t`, which should be bound at de-Bruijin index 1, since it was the argument to the outer lambda.

In non-de-Bruijin indicies, the dependent identity should be:
`\t : * 0 -> \x : t -> (x : t)` with type `pi (t : * 0) . pi (x : t) . t`.
That term should translate to de-Bruijin as `\.\. (0 : 1)`, right?

Do you see my confusion?

Perhaps it would be helpful to list out the indices when type-checking different sub-terms.

Reply:
> In non-de-Bruijin indicies, the dependent identity should be:
> `\t : * 0 -> \x : t -> (x : t)` with type `pi (t : * 0) . pi (x : t) . t`.
> That term should translate to de-Bruijin as `\.\. (0 : 1)`, right? That's what you have on line 55. So why is the type on line 77 de-Bruijin index 0 (`x`) instead of de-Bruijin index 1 (`t`)?
> Do you see my confusion?

I see. Line 62 is the annotation for line 61. It says that the type of the output (`x`) is the first input.

For the annotation on lines 67-77, the type checker knows that for the identity function, the output's type has to be the same as the second input. (Hence on line 77, it's index 0) But that's fine because the type of `x` is bounded to the first input's value (on line 76).

Did that answer your question?

Reply:
It seems the type checker as is annotates such that the return value takes the input as is, not the input's value. I.e., on L77, it takes in the second input, `Nat`, although on L76, it takes in the first input's value.

Will add the following tests to experiment further: 

\t1.\t2.\(x : t1).x
\t1.\t2.\(x : t2).x
\t.\(x:t).\(y:t).x
and \t.\(x:t).\(y:t).y

Will also look into other languages to see if this is the correct/desired type checker behaviour.

Reply:
I have a better way of explaining why it is correct that the return is the type of `x`, not `x`:

In the annotation all we ever take in is the type of `x`, i.e., `t`. We do not take in the value of `x`. The second inputs is annotated as the first input (i.e., the type of `x`, `t`). And the return type is bound to the second input's value, in this case `t`.

Reply:
The annotation doesn't contain the value of `x`, yes - it doesn't contain the value of `t` either - but it should express that the identity function should take *some* t as the first input, an x of type *t* as the second input, and return something of type *t*; it can refer to the values in the abstract (then the typechecker will fill them in when we check the annotation against an actual term). The only thing the annotation can refer to are the values of previous inputs (even though it doesn't know what they are yet) - this is how the value of *t* is referred to when specifying the type of the argument `x` - since values, not types, are substituted in when performing application:

![2019-11-27-080659_761x268_scrot](https://user-images.githubusercontent.com/24760279/69701300-ea73b800-10ec-11ea-9922-4e7e1bccba74.png)


The value of the first input, `t`, is the type of the second input and the type of the return value. We don't want to refer to "the type of x" anywhere, annotations shouldn't bind types of anything (when we check the annotation, we should check that the type of the body - `(x : t)` - is `t`, which it should be).

Reply:
So, this does pass:

```haskell
depIdentityCompTy ∷ ∀ primTy primVal. IR.Annotation primTy primVal
depIdentityCompTy =
  ( SNat 0,
    IR.VPi
      (SNat 0)
      (IR.VStar 0) -- first input is of type type
      (\t ->
          ( IR.VPi
              (SNat 1)
              (IR.VNeutral (IR.NFree (IR.Local 0))) -- second input is of type of the first input
              (const t)
          )   
      )   
  )
```

Why do we have both the lambda arguments and the bound indices? It seems like only one should be necessary.

Reply:
I think @andy-morris looked at this and also thought an index was off, but I'll let him explain his analysis.

Reply:
> 
> ```haskell
> depIdentityCompTy ∷ ∀ primTy primVal. IR.Annotation primTy primVal
> depIdentityCompTy =
>   ( SNat 0,
>     IR.VPi
>       (SNat 0)
>       (IR.VStar 0) -- first input is of type type
>       (\t ->
>           ( IR.VPi
>               (SNat 1)
>               t
>               (const t)
>           )   
>       )   
>   )
> ```
> 
> And this fails, which I would also expect:
> 
> ```haskell
> depIdentityCompTy ∷ ∀ primTy primVal. IR.Annotation primTy primVal
> depIdentityCompTy =
>   ( SNat 0,
>     IR.VPi
>       (SNat 0)
>       (IR.VStar 0) -- first input is of type type
>       (\t ->
>           ( IR.VPi
>               (SNat 1)
>               t
>               (\x -> x)
>           )   
>       )   
>   )
> ```
> 

In general I don't think we should mix Haskell's `\x` function into my type checker DSL. The above inputs would never be passed in in practice because the type checker wouldn't generate such annotations.

Reply:
> In general I don't think we should mix Haskell's \x function into my type checker DSL. The above inputs would never be passed in in practice because the type checker wouldn't generate such annotations.

Hmm, are you sure? HOAS is used in value evaluation, e.g. [here](https://github.com/cryptiumlabs/juvix/blob/develop/src/Juvix/Core/IR/Evaluator.hs#L27).

Reply:
> I think @andy-morris looked at this and also thought an index was off, but I'll let him explain his analysis.

I'm not sure this little scribble will be that enlightening, but yeah, I'm pretty sure that the type should end in index `1` rather than `0`, since it has to skip over the binder for `x` itself.

(I thought I'd already posted this but it doesn't seem to have fully gone through, sorry about that)

![20191129_185156](https://user-images.githubusercontent.com/469943/70076126-6cd50e00-15fe-11ea-89a3-c0337f565554.jpg)

Reply:
Nice! Type theory in color! I don't think I understand - but it'll be easier to discuss after the logging is finished - we'll do that then. See #223

Reply:
> Hmm, are you sure? HOAS is used in value evaluation, e.g. [here](https://github.com/cryptiumlabs/juvix/blob/develop/src/Juvix/Core/IR/Evaluator.hs#L27).
> 
> I would be fine with eliminating it, but I think we should do so uniformly if we do.

You're right, `VLam` and `VPi` take functions. Perhaps the annotations should be written in that form actually.